simpeg/directives/_regularization.py

```diff
@@ -0,0 +1,482 @@
+from typing import Literal
+from .directives import InversionDirective, UpdatePreconditioner, BetaSchedule
+from simpeg.regularization import Sparse, BaseSparse, SmoothnessFirstOrder
+from simpeg.utils import validate_integer, validate_float
+from simpeg.utils.code_utils import deprecate_property
+
+
+@dataclass
+class IRLSMetrics:
+    """
+    Data class to store metrics used by the IRLS algorithm.
+
+    Parameters
+    ----------
+
+    input_norms: List of norms temporarily stored during the initialization.
+    irls_iteration_count: Number of IRLS iterations.
+    start_irls_iter: Iteration number when the IRLS process started.
+    f_old: Previous value of the regularization function.
+    mode: Mode of the IRLS algorithm, 1 for l2-norm and 2 for lp-norm iterations.
+    """
+
+    input_norms: list[float]
+    irls_iteration_count = 0
+    start_irls_iter = 0
+    f_old: float = 0.0
+    mode: Literal[1, 2] = 1
+
+
+class UpdateIRLS(InversionDirective):
+    """
+    Directive to control the IRLS iterations for regularization.Sparse.
+
+    Parameters
+    ----------
+
+    beta_search: bool
+        Proceed with a beta search step if the target misfit is outside the tolerance.
+    misfit_tolerance: float
+        Tolerance for the target misfit.
+    chifact_start: float
+        Starting chi factor for the IRLS iterations.
+    chifact_target: float
+        Target chi factor for the IRLS iterations.
+    cooling_factor: float
+        Factor to cool the IRLS threshold epsilon.
+    f_min_change: float
+        Minimum change in the regularization function to continue the IRLS iterations.
+    max_irls_iterations: int
+        Maximum number of IRLS iterations.
+    percentile: float
+        Percentile of the function values used to determine the initial IRLS threshold.
+    """
+
+    def __init__(
+        self,
+        beta_schedule: BetaSchedule | None = None,
+        chifact_start: float = 1.0,
+        chifact_target: float = 1.0,
+        cooling_factor: float = 1.2,
+        f_min_change: float = 1e-2,
+        max_irls_iterations: int = 20,
+        misfit_tolerance: float = 1e-1,
+        percentile: int = 100,
+        **kwargs,
+    ):
+        self._metrics: IRLSMetrics | None = None
+        self.beta_schedule: BetaSchedule = beta_schedule
+        self.chifact_start: float = chifact_start
+        self.chifact_target: float = chifact_target
+        self.cooling_factor: float = cooling_factor
+        self.f_min_change: float = f_min_change
+        self.max_irls_iterations: int = max_irls_iterations
+        self.misfit_tolerance: float = misfit_tolerance
+        self.percentile: int = percentile
+
+        if "silent" in kwargs:
+            warnings.warn(
+                "The `silent` keyword argument is deprecated. "
+                "Use `verbose` instead.",
+                FutureWarning,
+                stacklevel=2,
+            )
+            kwargs["verbose"] = not kwargs.pop("silent")
```

Comment:
Similar thing with the arguments. I don't think we should support a deprecated `silent` argument. Let's just keep with `verbose`. BTW, would be nice to include `verbose` as an optional argument in the constructor (and include it in the docstring), so it's more discoverable than having to pass it through kwargs.